### PR TITLE
[mlir][spirv] Add tests for `scf.while` and `scf.for` in `convert-to-spirv` pass

### PR DESCRIPTION
**Pull request by obtuseangleAZ**
This PR adds lit tests that check for `scf.while` and `scf.for` conversions for the `convert-to-spirv` pass, introduced in #95942.

### CODE REVIEW

**Review by kuhar — COMMENTED** [review]
Are these tested by the scf to spirv conversion? If yes, could we trim these tests down so that we don't duplicate tests?

**Comment by angelz913** [comment]
> Are these tested by the scf to spirv conversion? If yes, could we trim these tests down so that we don't duplicate tests?

Yes, these tests were mostly from the `scf-to-spirv` conversion. I just realized that I forgot to include them in the initial version PR, which made me think that `scf.while` and `scf.for` were not covered. To prevent duplicate tests, I can try to include tests that combine with memref (with storage class conversion that is not covered in `scf-to-spirv`).

**Comment by kuhar** [comment]
I think we can check that they produce `spirv.mlir.loop`, `spirv.Branch`, etc., but not the exact instruction sequence. This way we will know that the conversion produced something reasonable, end the exact expansion is tested in the scf to spirv test.

File: mlir/test/Conversion/ConvertToSPIRV/scf.mlir [hunk 1]
```diff
@@ -33,6 +33,17 @@ func.func @if_yield(%arg0: i1) -> f32 {
 }
 
 // CHECK-LABEL: @while
+// CHECK:       spirv.mlir.loop {
+// CHECK:         spirv.Branch ^[[HEADER:.*]](%{{.*}} : i32)
+// CHECK:       ^[[HEADER]]
+// CHECK:         spirv.Store "Function"
```
Comment (kuhar) [inline mlir/test/Conversion/ConvertToSPIRV/scf.mlir:39]: Why is this store important?

File: mlir/test/Conversion/ConvertToSPIRV/scf.mlir [hunk 2]
```diff
@@ -33,6 +33,17 @@ func.func @if_yield(%arg0: i1) -> f32 {
...
+// CHECK:         spirv.BranchConditional %{{.*}}, ^[[BODY:.*]](%{{.*}} : i32), ^[[MERGE:.*]]
+// CHECK:       ^[[BODY]]
+// CHECK:       spirv.Branch
+// CHECK:       ^[[MERGE]]
+// CHECK:         spirv.mlir.merge
+// CHECK:       }
+// CHECK:       spirv.Load "Function"
```
Comment (kuhar) [inline mlir/test/Conversion/ConvertToSPIRV/scf.mlir:46]: Also this load

File: mlir/test/Conversion/ConvertToSPIRV/scf.mlir [hunk 3]
```diff
@@ -45,3 +56,30 @@ func.func @while(%arg0: i32, %arg1: i32) -> i32 {
...
+// CHECK:         spirv.BranchConditional %{{.*}}, ^[[BODY:.*]], ^[[MERGE:.*]]
+// CHECK:       ^[[BODY]]
+// CHECK-DAG:     spirv.Store "Function"
+// CHECK-DAG:     spirv.Store "Function"
+// CHECK:       spirv.Branch ^[[HEADER]]
+// CHECK:       ^[[MERGE]]
+// CHECK:         spirv.mlir.merge
+// CHECK:      }
+// CHECK-DAG:  spirv.Load "Function"
+// CHECK-DAG:  spirv.Load "Function"
```
Comment (kuhar) [inline mlir/test/Conversion/ConvertToSPIRV/scf.mlir:73]: Same here

File: mlir/test/Conversion/ConvertToSPIRV/scf.mlir (hunk 1 above)
Comment (angelz913) [reply mlir/test/Conversion/ConvertToSPIRV/scf.mlir:39]: I think that when we lower a `scf.while` with `scf.yield`, the header block contains `spirv.Store "Function"`, so I included this to ensure that the `while` is lowered properly.

File: mlir/test/Conversion/ConvertToSPIRV/scf.mlir (hunk 2 above)
Comment (angelz913) [reply mlir/test/Conversion/ConvertToSPIRV/scf.mlir:46]: Outside the `spirv.mlir.loop`, we should have a  `spirv.Load "Function"` to call the previously stored function.

File: mlir/test/Conversion/ConvertToSPIRV/scf.mlir (hunk 3 above)
Comment (angelz913) [reply mlir/test/Conversion/ConvertToSPIRV/scf.mlir:73]: This is similar to the `scf.while` case, except that the `scf.yield` is producing two results, which corresponds to two `spirv.Store "Function"` and `spirv.Load "Function"` pairs.

File: mlir/test/Conversion/ConvertToSPIRV/scf.mlir (hunk 1 above)
Comment (kuhar) [reply mlir/test/Conversion/ConvertToSPIRV/scf.mlir:39]: Yes, but I'd expect this to be tested by scf to spirv tests, not convert to spirv

Comment (angelz913) [reply mlir/test/Conversion/ConvertToSPIRV/scf.mlir:39]: I see. I can remove them if they are redundant.

**Review by kuhar — APPROVED** [review]
LGTM

**Pull request merged**